pilot/pkg/networking/plugin/plugin.go

```diff
@@ -16,24 +16,66 @@ package plugin
 
 import (
 	xdsapi "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	"github.com/envoyproxy/go-control-plane/envoy/api/v2/listener"
+	http_conn "github.com/envoyproxy/go-control-plane/envoy/config/filter/network/http_connection_manager/v2"
 
 	"istio.io/istio/pilot/pkg/model"
 )
 
-// Callbacks represents the interfaces implemented by code that modifies the default output of
-// networking. Examples include AuthenticationPlugin that sets up mTLS authentication on the inbound Listener
+// ListenerType is the type of listener.
+type ListenerType int
+
+const (
+	// ListenerTypeTCP is a TCP listener.
+	ListenerTypeTCP = iota
+	// ListenerTypeHTTP is an HTTP listener.
+	ListenerTypeHTTP
+)
+
+// CallbackListenerInputParams is a set of values passed to On*Listener callbacks. Not all fields are guaranteed to
+// be set, it's up to the callee to validate required fields are set and emit error if they are not.
+// These are for reading only and should not be modified.
+type CallbackListenerInputParams struct {
+	// ListenerType is the type of listener (TCP, HTTP etc.)
+	ListenerType ListenerType
+	// Env is the model environment.
+	Env *model.Environment
+	// Node is the node the listener is for.
+	Node *model.Proxy
+	// ProxyInstances is a slice of all proxy service instances in the mesh.
+	ProxyInstances []*model.ServiceInstance
+	// ServiceInstance is the service instance colocated with the listener (applies to sidecar).
+	ServiceInstance *model.ServiceInstance
+	// ServicePort is the port for the service
+	// TODO(mostrowski): can get from ServiceInstance?
+	ServicePort *model.Port
+}
+
+// CallbackListenerMutableObjects is a set of objects passed to On*Listener callbacks. Fields may be nil or empty.
+// Any lists should not be overridden, but rather only appended to.
+// Non-list fields may be mutated; however it's not recommended to do this since it can affect other plugins in the
+// chain in unpredictable ways.
+type CallbackListenerMutableObjects struct {
+	// HTTPFilters is the slice of HTTP filters for the Listener. Append to only.
+	HTTPFilters []*http_conn.HttpFilter
+	// TCPFilters is the slice of TCP filters for the Listener. Append to only.
+	TCPFilters []listener.Filter
+	// Listener is the listener being built. Any field may be modified, but changing other than appending to slices
+	// is discouraged.
+	Listener *xdsapi.Listener
+}
+
+// ListenerPlugin is a plugin called during the construction of a xdsapi.Listener which may alter the Listener in any
+// way. Examples include AuthenticationPlugin that sets up mTLS authentication on the inbound Listener
 // and outbound Cluster, the mixer plugin that sets up policy checks on the inbound listener, etc.
-type Callbacks interface {
+type ListenerPlugin interface {
```

Comment:
Why rename this? this has non listener stuff as well, like clusters and routes

Reply:
i was struggling to come up with a better name. callbacks doesn't sound right... how about just Plugin?

Reply:
Sure. @ZackButcher changed it to callback. Plugin sounds fine. Just not a listenerplugin

Reply:
done

Reply:
No, I explicitly renamed this away from plugin. `plugin.Plugin` is super redundant. Please put this back.